wallet-new/integration/Functions.hs

```diff
@@ -21,467 +25,556 @@ import           Cardano.Wallet.API.V1.Types (Account (..), AccountIndex, Accoun
                                               PaymentDistribution (..), PaymentSource (..),
                                               SpendingPassword, Transaction (..), V1 (..),
                                               Wallet (..), WalletAddress (..), WalletId,
-                                              WalletOperation (..), WalletUpdate (..))
+                                              WalletOperation (..), WalletUpdate (..), unV1)
 
 import           Cardano.Wallet.API.V1.Migration.Types (migrate)
-import           Cardano.Wallet.Client (ClientError (..), WalletClient (..), getAccounts,
-                                        getAddressIndex, getTransactionIndex, getWallets)
+import           Cardano.Wallet.Client (ClientError (..), Response (..), ServantError (..),
+                                        WalletClient (..), getAccounts, getAddressIndex,
+                                        getTransactionIndex, getWallets, hoistClient)
 
-import           Pos.Core (mkCoin)
-import           Pos.Util (maybeThrow)
+import           Pos.Core (getCoin, mkCoin)
 import qualified Pos.Wallet.Web.ClientTypes.Types as V0
 
 import           Error
 import           Types
 
-
--- | The top function that we need to run in order
--- to test the backend.
+newtype RefT s m a
```

Comment:
We were having a problem with state! It would roll back on an `empty` from IO. So I switched to an `IORef` variant that doesn't roll back. This has a custom `MonadState` implementation that writes the IORef.

Reply:
It is not obvious to me why this was happening (I am a dumb person). Would you give some more pointers (it seems important to understand)

Reply:
Aha, I got it (after some code review) ty

Reply:
Consider how `StateT` does an alternative:

```haskell
StateT f <|> StateT g =
  StateT $ \s -> 
    f s <|> g s
```

If `f` fails with empty, then we pass the same state to `g`. So if we have:

```haskell
foo = do
  bar
  baz
  empty

test = foo <|> quux
```

When we call `foo`, `bar` and `baz` both modify the state. However, `empty` cancels the computation, and `StateT`'s alternative instance passes the state to `quux` that was initially passed to `foo`.

If you layer the transformers differently (eg `MaybeT (StateT ...)`) then the backtracking is different. But `IO` must be on the bottom.